### PR TITLE
HDFS-14869 Copy renamed files which are not excluded anymore by filter

### PR DESCRIPTION
**Pull request by aasha**

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 124 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1354 | trunk passed |
| +1 | compile | 25 | trunk passed |
| +1 | checkstyle | 20 | trunk passed |
| +1 | mvnsite | 28 | trunk passed |
| +1 | shadedclient | 837 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 21 | trunk passed |
| 0 | spotbugs | 41 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 40 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 25 | the patch passed |
| +1 | compile | 19 | the patch passed |
| +1 | javac | 19 | the patch passed |
| -0 | checkstyle | 15 | hadoop-tools/hadoop-distcp: The patch generated 4 new + 47 unchanged - 0 fixed = 51 total (was 47) |
| +1 | mvnsite | 22 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 855 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 19 | the patch passed |
| +1 | findbugs | 44 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 871 | hadoop-distcp in the patch passed. |
| +1 | asflicense | 44 | The patch does not generate ASF License warnings. |
| | | 4475 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1530/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1530 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux cb0035532a13 4.15.0-54-generic #58-Ubuntu SMP Mon Jun 24 10:55:24 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 7b6219a |
| Default Java | 1.8.0_222 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1530/1/artifact/out/diff-checkstyle-hadoop-tools_hadoop-distcp.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1530/1/testReport/ |
| Max. process+thread count | 347 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-distcp U: hadoop-tools/hadoop-distcp |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1530/1/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

File: hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/TestDistCpSync.java [hunk 1]
```diff
@@ -747,4 +754,84 @@ public void testSyncSnapshotTimeStampChecking() throws Exception {
...
+
+  private void generateFilterFile(String directory, String fileName){
+    File theDir = new File(directory);
+    boolean threwException = false;
+    if (!theDir.exists()) {
+      theDir.mkdir();
+    }
+    String str = ".*\\.staging.*";
+    BufferedWriter writer = null;
+    try {
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/TestDistCpSync.java]: use try-with resources

File: hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/TestDistCpSync.java [hunk 2]
```diff
@@ -747,4 +754,84 @@ public void testSyncSnapshotTimeStampChecking() throws Exception {
...
+      writer = new BufferedWriter(new FileWriter(directory + "/" + fileName));
+      writer.write(str);
+      writer.close();
+    } catch (IOException e) {
+      threwException = true;
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/TestDistCpSync.java]: don't catch, just throw and change the signature. Stack traces are too important to lose

File: hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/TestDistCpSync.java [hunk 3]
```diff
@@ -747,4 +754,84 @@ public void testSyncSnapshotTimeStampChecking() throws Exception {
...
+  }
+
+  private void deleteFilterFile(String directory, String fileName) {
+    File theDir = new File(directory);
+    if (!theDir.exists()) {
+      return;
+    }
+    File file = new File(directory + "/" + fileName);
+    if (file.exists()) {
+      file.delete();
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/TestDistCpSync.java]: no need for the exists check, delete is no-op if nothing is there

File: hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/TestDistCpSync.java [hunk 4]
```diff
@@ -747,4 +754,84 @@ public void testSyncSnapshotTimeStampChecking() throws Exception {
...
+  }
+
+  @Test
+  public void testSync10() throws Exception {
+    try {
+      Path sourcePath = new Path(dfs.getWorkingDirectory(), "source");
+      initData10(sourcePath);
+      dfs.allowSnapshot(sourcePath);
+      dfs.createSnapshot(sourcePath, "s1");
+      generateFilterFile("/tmp", "filters.txt");
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/TestDistCpSync.java]: can't rely on /tmp, especially on jenkins systems. use the temp file API call in File itself,

File: hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/DistCpSync.java [hunk 5]
```diff
@@ -213,18 +216,30 @@ private boolean getAllDiffs() throws IOException {
...
+        final Path source =
+                new Path(DFSUtilClient.bytes2String(entry.getSourcePath()));
+        final Path relativeSource = new Path(Path.SEPARATOR + source);
         if (dt == SnapshotDiffReport.DiffType.MODIFY ||
-            dt == SnapshotDiffReport.DiffType.CREATE ||
-            dt == SnapshotDiffReport.DiffType.DELETE) {
-          final Path source =
-              new Path(DFSUtilClient.bytes2String(entry.getSourcePath()));
-          list.add(new DiffInfo(source, null, dt));
+                dt == SnapshotDiffReport.DiffType.CREATE ||
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/DistCpSync.java]: no need to change indentation

File: hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/DistCpSync.java [hunk 6]
```diff
@@ -213,18 +216,30 @@ private boolean getAllDiffs() throws IOException {
...
+                dt == SnapshotDiffReport.DiffType.DELETE) {
+          if (copyFilter.shouldCopy(relativeSource)) {
+            list.add(new DiffInfo(source, null, dt));
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/DistCpSync.java]: why remove the bytes2String? surely its there for a reason? removal could be a regression

File: hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/DistCpSync.java [hunk 7]
```diff
@@ -213,18 +216,30 @@ private boolean getAllDiffs() throws IOException {
...
-          final Path source =
-              new Path(DFSUtilClient.bytes2String(entry.getSourcePath()));
           final Path target =
-              new Path(DFSUtilClient.bytes2String(entry.getTargetPath()));
-          list.add(new DiffInfo(source, target, dt));
+                  new Path(DFSUtilClient.bytes2String(entry.getTargetPath()));
+          final Path relativeTarget = new Path(Path.SEPARATOR + target);
+          if (copyFilter.shouldCopy(relativeSource)) {
+            if (copyFilter.shouldCopy(relativeTarget)) {
+              list.add(new DiffInfo(source, target, dt));
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/DistCpSync.java]: again, why no bytes2String

**Comment by bshashikant** [comment]
Thanks @aasha for working on this. The changes overall look good to me. I have one comment.
 Can we also modify the test to include a case of rename, where the filter excludes the rename target but the source file is included which after the distcp completes should be deleted on the target?

**Comment by steveloughran** [comment]
LGTM, we just need yetus to see it. can you do a rebase + force push to see if that works?

**Comment by bshashikant** [comment]
@steveloughran , even an empty commit seems to not trigger yetus. Any other way to trigger this?
Can the patch be just uploaded into the jira to see if yetus picks up?

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m 12s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  21m  9s |  trunk passed  |
| +1 :green_heart: |  compile  |   0m 24s |  trunk passed  |
| +1 :green_heart: |  checkstyle  |   0m 20s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   0m 27s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  14m 55s |  branch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 21s |  trunk passed  |
| +0 :ok: |  spotbugs  |   0m 41s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   0m 39s |  trunk passed  |
| -0 :warning: |  patch  |   0m 54s |  Used diff version of patch file. Binary files and potentially other changes not applied. Please rebase and squash commits if necessary.  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 25s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 19s |  the patch passed  |
| +1 :green_heart: |  javac  |   0m 19s |  the patch passed  |
| -0 :warning: |  checkstyle  |   0m 14s |  hadoop-tools/hadoop-distcp: The patch generated 10 new + 47 unchanged - 0 fixed = 57 total (was 47)  |
| +1 :green_heart: |  mvnsite  |   0m 22s |  the patch passed  |
| -1 :x: |  whitespace  |   0m  0s |  The patch has 1 line(s) that end in whitespace. Use git apply --whitespace=fix <<patch_file>>. Refer https://git-scm.com/docs/git-apply  |
| +1 :green_heart: |  shadedclient  |  14m 52s |  patch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 18s |  the patch passed  |
| +1 :green_heart: |  findbugs  |   0m 43s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  12m 42s |  hadoop-distcp in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 28s |  The patch does not generate ASF License warnings.  |
|  |   |  71m 16s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.5 Server=19.03.5 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1530/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1530 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 8f01e99a911b 4.15.0-66-generic #75-Ubuntu SMP Tue Oct 1 05:24:09 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 47fdae7 |
| Default Java | 1.8.0_222 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1530/2/artifact/out/diff-checkstyle-hadoop-tools_hadoop-distcp.txt |
| whitespace | https://builds.apache.org/job/hadoop-multibranch/job/PR-1530/2/artifact/out/whitespace-eol.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1530/2/testReport/ |
| Max. process+thread count | 314 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-distcp U: hadoop-tools/hadoop-distcp |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1530/2/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   3m 26s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  22m 22s |  trunk passed  |
| +1 :green_heart: |  compile  |   0m 30s |  trunk passed  |
| +1 :green_heart: |  checkstyle  |   0m 27s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   0m 37s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  17m 30s |  branch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 26s |  trunk passed  |
| +0 :ok: |  spotbugs  |   0m 57s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   0m 55s |  trunk passed  |
| -0 :warning: |  patch  |   1m 16s |  Used diff version of patch file. Binary files and potentially other changes not applied. Please rebase and squash commits if necessary.  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 32s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 23s |  the patch passed  |
| +1 :green_heart: |  javac  |   0m 23s |  the patch passed  |
| -0 :warning: |  checkstyle  |   0m 19s |  hadoop-tools/hadoop-distcp: The patch generated 10 new + 47 unchanged - 0 fixed = 57 total (was 47)  |
| +1 :green_heart: |  mvnsite  |   0m 28s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  17m 55s |  patch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 22s |  the patch passed  |
| +1 :green_heart: |  findbugs  |   0m 55s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  16m 31s |  hadoop-distcp in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 36s |  The patch does not generate ASF License warnings.  |
|  |   |  86m  7s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.5 Server=19.03.5 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1530/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1530 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux fe4151e5f851 4.15.0-66-generic #75-Ubuntu SMP Tue Oct 1 05:24:09 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / da9aa9c |
| Default Java | 1.8.0_222 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1530/3/artifact/out/diff-checkstyle-hadoop-tools_hadoop-distcp.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1530/3/testReport/ |
| Max. process+thread count | 303 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-distcp U: hadoop-tools/hadoop-distcp |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1530/3/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

**Comment by bshashikant** [comment]
@aasha , can you please address the checkstyle issues?

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m  7s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  1s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  21m  6s |  trunk passed  |
| +1 :green_heart: |  compile  |   0m 24s |  trunk passed  |
| +1 :green_heart: |  checkstyle  |   0m 23s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   0m 27s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  14m 32s |  branch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 20s |  trunk passed  |
| +0 :ok: |  spotbugs  |   0m 45s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   0m 43s |  trunk passed  |
| -0 :warning: |  patch  |   1m  0s |  Used diff version of patch file. Binary files and potentially other changes not applied. Please rebase and squash commits if necessary.  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 24s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 19s |  the patch passed  |
| +1 :green_heart: |  javac  |   0m 19s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   0m 14s |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   0m 22s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  14m 43s |  patch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 18s |  the patch passed  |
| +1 :green_heart: |  findbugs  |   0m 43s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  12m 44s |  hadoop-distcp in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 28s |  The patch does not generate ASF License warnings.  |
|  |   |  70m 40s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.5 Server=19.03.5 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1530/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1530 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 8127bfe342ab 4.15.0-66-generic #75-Ubuntu SMP Tue Oct 1 05:24:09 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / e575df5 |
| Default Java | 1.8.0_222 |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1530/4/testReport/ |
| Max. process+thread count | 309 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-distcp U: hadoop-tools/hadoop-distcp |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1530/4/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

**Pull request merged**